Global Issues: Simplifying addition of new algorithms and Main Page issues

el-young opened the pull request:
- Reworked Main Menu to derive links directly from the master list.
- Removed repetitive code from `/mainmenu` and related `.scss` files.
- Added keyword search functionality to the search bar.
- Fixed scaling issues (design still provisional; will follow up on desired layout).

Note: We are aware `AlgorithmMenu.js` still needs modification, but there are some weird issues when attempting to import from `/algorithms`.

lee-naish commented:
Well done!

We need to set up proper communication channels between your group and the clients (plus other groups)!

I'm holding off merging for now due to the following:

For keywords, I think it would be much more convenient for future development if there was a keywords field of the elements of the allalgs list , eg 'isort': {  name: 'Insertion Sort', keywords = ["N^2", "adaptive"], ...

BTW, I tried searching for N^2 and nothing came up - not sure if this functionality is there yet or what the issue is.

Also, I didn't understand the comment about conflicting keywords.

Great that the main menu page now scrolls etc. We will discuss how it looks in more detail some time (I'm all for simplicity; Linda is more picky). I noticed that algorithm names that are too long and wrap around are not centered. Possibly best wait until further discussion with Linda unless there is a very quick fix. I tried changing the category of Gift Wrapping to Geometry and the main menu was updated very nicely.  However, the drop-down menu (on the pages with animations) didn't include the new category.

Good work - will see you soon.

lee-naish commented:
One other thing, we are planning to merge to 2025_sem2 for now, not dev.

el-young commented:
Noted.

Yes the algorithm menu changes were not pushed here, we’ve been blocked for a couple of days on the algorithm menu work and wanted to check whether it’s acceptable to ask for help at this stage. If it isn’t, please let me know and I’ll hold off. If it is, here’s the issue and what we’ve tried:

## Problem

Importing `DeployedAlgorithms` from `src/algorithms/index.js` inside `AlgorithmMenu.js` crashes the app with an error along the lines of “accessing an export before its value is initialized,”. Strangely, the same import line works in `MainMenu.js`  with no issue at all.

```js
// AlgorithmMenu.js
import React, { useState } from 'react';
import '../styles/AlgorithmMenu.scss';

// Get the base URL dynamically
const baseUrl = window.location.origin;

// TODO: I dont understand why this breaks everything, I used this import
// in MainMenu.js with no issue. 
// Circular imports (npx madge src/components/AlgorithmMenu.js --circular)
// but MainMenu.js (npx madge src/components/MainMenu.js --circular) 
// has the exact same circular imports???
// import { DeployedAlgorithms } from '../algorithms';

// This works but prod build doesnt allow top level await
// so we can not commit this.
const { DeployedAlgorithms } = await import('../algorithms');

// Dynamically create URL's in algorithm menu from master list (src/algorithms/index.js).
// Only the entries that have noDeploy property set to false are included on the menu.
const algorithms = Object.fromEntries(
  DeployedAlgorithms.map(({ category, algorithms }) => [
    category,
    Object.fromEntries(
      algorithms.map(({ name, shorthand, mode }) => [
        name,
        `${baseUrl}/?alg=${shorthand}&mode=${mode}`,
      ])
    ),
  ])
);

function AlgorithmMenu() {
  const [isOpen, setIsOpen] = useState(false);
  const [activeCategory, setActiveCategory] = useState(null);

  return (
    <div className="algorithm-menu">
      <button onClick={() => setIsOpen(!isOpen)}>Menu</button>
      {isOpen && (
        <div className="dropdown">
          {Object.entries(algorithms).map(([category, algs]) => (
            <div
              key={category}
              className="category"
              onMouseEnter={() => setActiveCategory(category)}
              onMouseLeave={() => setActiveCategory(null)}
            >
              {category}
              {activeCategory === category && (
                <div className="subcategory">
                  {Object.entries(algs).map(([alg, url]) => (
                    <a key={alg} href={url || '#'}>
                      {alg}
                    </a>
                  ))}
                </div>
              )}
            </div>
          ))}
        </div>
      )}
    </div>
  );
}

export default AlgorithmMenu;
```

This will make the website run as expected but it uses a hacky and brittle `await` at top level which violates the commit rules anyway so its uncommitable. 

The team and I are JS newbies, so we did our research but to no avail, is there a simple reason why this import line in this specific file breaks the entire site but the same import line in `MainMenu.js` does not?

lee-naish commented:
It's fine to ask for help. I'm not sure if I will be able to help much as I don't have in-depth knowledge of JS either - I've just learned a bit as I've gone along but things like await are still black magic to me and I've been unable to get any of those spells working. I'll let you know if I feel bombarded by emails I would rather not have.

In this case I don't really know whats going on, but I think I may be able to help. I think you may have run into an annoying npm bug that fails to recompile everything properly when some files are modified. I've noticed this occasionally when "impossible"/weird error message pop up during compilation. Try doing a fresh recompile/install.  I added

import { DeployedAlgorithms } from '../algorithms';

to AlgorithmMenu.js and there were no compile-time errors (DeployedAlgorithms isn't quite the same format as algorithms in that file so I didn't actually get working code but close enough to be helpful I think). Fingers crossed...

el-young commented:
Yeah, it definitely seems like a niche issue. No worries, I just wanted to make sure we weren’t missing anything obvious.

el-young commented:
Also when you have the time, can you check the latest version of the fork.

- We added a CLI for the `addNewAlgorithm.js` file for now its called `addNewAlgorithmScript.js`
- You mentioned in the original file that it would be nice to have 
   - verification checks on user input (this is yet to be done but we plan on doing so)
   - choice of what default algorithm code is used. This proves tricky to do because the prefixes are not consistent for files
       - For example with heapsort we have `src/algorithms/controllers/heapSort.js` and `src/algorithms/extra-info/HSInfo.md` if this was `src/algorithms/extra-info/heapSortInfo.md` it would totally be doable, is it okay if we change the filenames and make sure the bits of code it will effect work?
- Do you approve of the CLI?
- Let us know if you want any extensions if so.

Note: I added an example at the bottom of the file, also if you want to see the link in the algorithm menu and not just the main menu appear when you use this script please go to `/src/components/AlgorithmMenu.js` and comment out the hard coded links and uncomment the await hack we discussed since thats the only way we have it working for now until I see the solution you found.

Also its yet to be checked on a Windows device, it should work since the library used states its cross compatible but I realise now it might not work with `\r\n` and `\n` quirks, so if you have a windows device the master list entry may not be created but all the other files will be.
